### ISSUE TITLE
Proposal: Compiler-Solved TypeName Constraint

### ISSUE THREAD

**Issue by eric-corumdigital** [purescript/purescript#3787]
```
class TypeName (a ∷ Type) (name ∷ Symbol) | a → name, name → a
```

For all types there is an instance of TypeName which bijectively maps types to their fully-qualified name.

A higher-order example is:

```
TypeName (Maybe Int) "(Data.Maybe.Maybe Prim.Int)"
```

Two reasons I know of to have this.

1) Dynamic.

```
foreign import data Any ∷ Type

data Dynamic = Dynamic String Any

toDynamic ∷ ∀ a name. IsSymbol name ⇒ TypeName a name ⇒ a → Dynamic
toDynamic x = Dynamic (reflectSymbol (SProxy∷_ name)) (unsafeCoerce x)

fromDynamic ∷ ∀ a name. IsSymbol name ⇒ TypeName a name ⇒ Dynamic → Maybe a
fromDynamic (Dynamic name x)
  | name == reflectSymbol (SProxy∷_ name) = Just (unsafeCoerce x)
  | otherwise = Nothing
```

2) Efficient type-level Type sets (because rows already have bespoke solvers). Every label is the type name of its type. For example:

```
( "(Data.Maybe.Maybe Prim.Int)" ∷ Maybe Int, "Prim.Int" ∷ Int )
```

**Comment by hdgarrood** [comment]
As far as I can tell this is basically Typeable; see also #4097 where we are leaning towards not adding the feature.

**Comment by rhendric** [comment]
I think this subsumes Typeable, but may have applications for which Typeable wouldn't suffice? Not saying I'm sold on it, but having the capability to manipulate symbols that represent (a subset of?) types would surely be a boon to type-level programming.

**Comment by JordanMartinez** [comment]
Would any opinions change if only the `a -> name` functional dependency existed? That alone could help with custom debug messages / errors, right?

**Comment by hdgarrood** [comment]
Not really. I'm more concerned that this feature would not pay for itself in terms of maintenance costs, even without the bidirectional functional dependencies, and that it could be a good way for bugs to sneak in; for example, you should not be able to observe type variable names, but with a class like this we'd have to be quite careful to avoid accidentally breaking that property.
